apollo-router/src/plugins/file_uploads/mod.rs

```diff
@@ -0,0 +1,644 @@
+        if !map_field.is_empty() {
+            map_field.sort_by_cached_key(|file, _| files_order.get_index_of(file));
+            req.subgraph_request
+                .extensions_mut()
+                .insert(SubgraphHttpRequestExtensions {
+                    multipart,
+                    map_field,
+                });
+        }
+    }
+    req
+}
+
+struct SubgraphHttpRequestExtensions {
+    multipart: MultipartRequest,
+    map_field: MapField,
+}
+
+const APOLLO_REQUIRE_PREFLIGHT: HeaderName = HeaderName::from_static("apollo-require-preflight");
+const TRUE: http::HeaderValue = HeaderValue::from_static("true");
+
+pub(crate) async fn http_request_wrapper(
+    mut req: http::Request<hyper::Body>,
+) -> http::Request<hyper::Body> {
+    let supergraph_result = req.extensions_mut().remove();
+    if let Some(supergraph_result) = supergraph_result {
+        let SubgraphHttpRequestExtensions {
+            multipart,
+            map_field,
+        } = supergraph_result;
+
+        let (mut request_parts, operations) = req.into_parts();
+        let form = MultipartFormData::new(operations, map_field, multipart);
+        request_parts
+            .headers
+            .insert(CONTENT_TYPE, form.content_type());
+        request_parts.headers.insert(APOLLO_REQUIRE_PREFLIGHT, TRUE);
+        return http::Request::from_parts(
+            request_parts,
+            hyper::Body::wrap_stream(form.into_stream().await),
+        );
+    }
+    req
+}
+
+#[derive(Clone)]
+struct MultipartRequest {
+    state: Arc<Mutex<MultipartRequestState>>,
+}
+
+struct MultipartRequestState {
+    multer: multer::Multipart<'static>,
+    limits: MultipartRequestLimits,
+    read_files_counter: usize,
+    file_sizes: Vec<usize>,
+    max_files_exceeded: bool,
+    max_files_size_exceeded: bool,
+}
+
+impl Drop for MultipartRequestState {
+    fn drop(&mut self) {
+        u64_counter!(
+            "apollo.router.operations.file_uploads",
+            "file uploads",
+            1,
+            "file_uploads.limits.max_file_size.exceeded" = self.max_files_size_exceeded,
+            "file_uploads.limits.max_files.exceeded" = self.max_files_exceeded
+        );
+
+        for file_size in &self.file_sizes {
+            u64_histogram!(
+                "apollo.router.operations.file_uploads.file_size",
+                "file upload sizes",
+                (*file_size) as u64
+            );
+        }
+        u64_histogram!(
+            "apollo.router.operations.file_uploads.files",
+            "number of files per request",
+            self.read_files_counter as u64
+        );
+    }
+}
+
+impl MultipartRequest {
+    fn new(request_body: hyper::Body, boundary: String, limits: MultipartRequestLimits) -> Self {
+        let multer = Multipart::with_constraints(
+            request_body,
+            boundary,
+            Constraints::new().size_limit(
+                SizeLimit::new().for_field("map", 10 * 1024), // hardcoded to 10kb
```

Comment:
NIT: If it's a hardcoded limit, maybe it would make sense to put it in a constant at the beginning of the file with a comment

Reply:
3a141e6